{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "84e9c5f7-344e-4cbb-aedd-6476157ccf86",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Workflow sections\n",
    "[1. How to use this notebook](#overview)\n",
    "\n",
    "[2. anacapa pipeline](#pipelineana)\n",
    "\n",
    "[3. Copying your data to Jupyter](#dataana)\n",
    "\n",
    "[4. Data overview](#spana)\n",
    "\n",
    "[5. Filtering taxonomic assignments](./anacapa_5_filtration.ipynb)\n",
    "\n",
    "[6. Alpha diversity](./anacapa_6_AD.ipynb)\n",
    "\n",
    "[7. Beta diversity](./anacapa_7_BD.ipynb)\n",
    "\n",
    "[8. Community Structure](./anacapa_8_CS.ipynb)\n",
    "\n",
    "[9. Differential abundance](./anacapa_9_DA.ipynb)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9e57b59-29a8-4ad1-b58a-d895cee07d7a",
   "metadata": {},
   "source": [
    "******************************"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f120af35-54eb-445d-8235-f0246a4c3655",
   "metadata": {},
   "source": [
    "# 1. How to use this notebook <a class=\"anchor\" id=\"overview\"></a>\n",
    "\n",
    "This is a [Jupyter notebook](https://jupyter.org/) for analysing 16S amplicon sequence data that has been processed by the [anacapa](https://ucedna.com/software) pipeline.\n",
    "\n",
    "Technically, however, as long as you have a table of taxa abundance and a samples tables, you can run this workflow (with some minor modifications)\n",
    "\n",
    "The majority of the sections in this notebook can be opened as separate Jupyter notebooks (by clicking on the link in the table of contents) and are designed to be run completely independantly.\n",
    "\n",
    "Jupyter notebooks are interactive documents that contain 'live code', which allows the user to complete an analysis by running code 'cells', which can be modified, updated or added to by the user.\n",
    "\n",
    "Individual Jupyter notebooks are based on a specific 'kernel', or analysis envirnment (mostly programming languages). This particular notebook is based on R. To see which version of R this notebook is based on, and as an example of running a code cell, click on the cell below and press the 'Run' button (top of the page).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18242159-e40b-49c5-882a-5ca6a1f02ea4",
   "metadata": {},
   "outputs": [],
   "source": [
    "R.Version()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7d101bd-0cec-4b85-be87-81658c25ab24",
   "metadata": {
    "tags": []
   },
   "source": [
    "You should now see details about the version of R installed for this notebook. Every other code cell can be run similarly. There are two main types of code cells, plain R code (as seen above) and [markdown](https://www.rstudio.com/wp-content/uploads/2015/02/rmarkdown-cheatsheet.pdf). \n",
    "\n",
    "Markdown is a simple language for formatting text and the instructions (including this cell), headings, etc are written in markdown code cells. See: https://www.rstudio.com/wp-content/uploads/2015/02/rmarkdown-cheatsheet.pdf. You can add additional cells (markdown or R) by clicking on the plus sign, and then in the dropdown box, selecting 'Markdown' or 'code'. This way you can add your own analysis code cells or your own notes in markdown."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac973d07-4950-47b4-9f19-16227c76e6b6",
   "metadata": {},
   "source": [
    "***********************************"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "730b735d-968d-4f9b-b9a9-dfb62c9c43fa",
   "metadata": {
    "tags": []
   },
   "source": [
    "# 2. anacapa pipeline <a class=\"anchor\" id=\"pipelineana\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd612541-3d39-4105-be9e-82e0aa43fc4b",
   "metadata": {},
   "source": [
    "This Jupyter eDNA analysis workflow has been designed to process data generated by the anacapa pipeline (though it can be readily modified to use output from other upstream eDNA analysis pipelines). Output data generated by anacapa can be directly used by this Jupyter workflow. Anacapa overview:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f24b95b-d45b-4ff0-8643-09d9f52bbe2e",
   "metadata": {},
   "source": [
    "> Anacapa is an eDNA toolkit that allows users to build comprehensive reference databases and assign taxonomy to raw multilocus metabarcode sequence data. It address longstanding needs of the eDNA for modular informatics tools, comprehensive and customizable reference databases, flexibility across high-throughput sequencing platforms, fast multilocus metabarcode processing, and accurate taxonomic assignment. Anacapa toolkit processes eDNA reads and assigns taxonomy using existing software or modifications to existing software. This modular toolkit is designed to analyze multiple samples and metabarcodes simultaneously from any Ilumina sequencing platform. A significant advantage of the Anacapa toolkit is that it does not require that paired reads overlap, or that both reads in a pair pass QC. Taxonomy results are generated for all read types and the user can decide which read types they wish to retain for downstream analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26af6f2e-064b-4293-b17a-3771c2a3b392",
   "metadata": {},
   "source": [
    "See the anacapa website for details: https://github.com/limey-bean/Anacapa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad5d3183-ca66-44a8-9884-a007b5e3e67b",
   "metadata": {},
   "source": [
    "Anacapa includes four modules:\n",
    "\n",
    "1. building reference libraries using CRUX\n",
    "2. running quality control (QC) and assigning Amplicon Sequence Variants (ASV) using Dada2 (Sequence QC and ASV Parsing),\n",
    "3. assigning taxonomy using Bowtie 2 and a Bowtie 2 specific Bayesian Least Common Ancestor (BLCA) (Assignment) and\n",
    "4. Running exploratory data analysis and generating ecological diversity summary statistics for a set of samples using ranacapa."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "841f4b0d-4204-4b4c-a5b6-c2b9946f2dc4",
   "metadata": {},
   "source": [
    "### Sequence QC and ASV Parsing using dada2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb5b65c9-4728-4aec-bef9-87121f7ffcdf",
   "metadata": {},
   "source": [
    "![](https://raw.githubusercontent.com/limey-bean/Anacapa/New-Master/dada2_QC_flow.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4404f5ec-3f86-40a9-9c59-9edf9d799a25",
   "metadata": {},
   "source": [
    "### Taxonomic Assignment using Bowtie 2 and BLCA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb3994ac-4e9d-46f3-93c3-f67a4cffba89",
   "metadata": {},
   "source": [
    "![](https://raw.githubusercontent.com/limey-bean/Anacapa/New-Master/Anacapa_class_flow.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b6d7547-e835-4525-968f-965b3f454d1f",
   "metadata": {},
   "source": [
    "The analysis in these Jupiter notebooks is for downstream analysis of eDNA data, based on a table of read/sequence counts per taxa, per sample. Thus, **to be able to run these Jupyter notebooks, you must have already run your sequence data through an upstream eDNA data processing pipeline, such as anacapa**."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5045d51c-d961-4f51-834b-de7d01cdc402",
   "metadata": {},
   "source": [
    "**CITING anacapa**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76495f74-fc5d-4870-ae35-3b8bd3e79e6b",
   "metadata": {},
   "source": [
    ">Curd, Emily E., Zack Gold, Gaurav S. Kandlikar, Jesse Gomer, Max Ogden, Taylor O'Connell, Lenore Pipes et al. \"Anacapa Toolkit: an environmental DNA toolkit for processing multilocus metabarcode datasets.\" Methods in Ecology and Evolution 10, no. 9 (2019): 1469-1475."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7cf2e3b-8167-4fbd-ba8e-9396cb6843b2",
   "metadata": {},
   "source": [
    "## R analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59507b1d-e633-4101-9b00-fe32b7f93de3",
   "metadata": {},
   "source": [
    "This Jupyter notebook, and the eDNA analysis therein, is based on R. Run the following code cell to see the version of R installed on this notebook:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac7fd02f-4d70-48a2-8687-d1c0a397476c",
   "metadata": {},
   "outputs": [],
   "source": [
    "R.Version()$version.string"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec3aab93-8d7f-4c58-9593-7cce51a46da6",
   "metadata": {},
   "source": [
    "The key R analysis package that we will use in this Jupyter Notebook is **[ampvis2 ('Tools for visualising amplicon data'](https://github.com/KasperSkytte/ampvis2))**. Output from the anacapa pipeline is used as input for ampvis2. Alpha diversity, beta diversity and community structure were mostly analysed using the ampvis2 package. Details of this analysis are in those sections of this analyis workflow.\n",
    "\n",
    "The final section of the workflow examines differential abundance of taxa between sample groups. The main R package used in this analysis is **[AMCOM-BC](https://bioconductor.org/packages/release/bioc/vignettes/ANCOMBC/inst/doc/ANCOMBC.html)**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48577c2f-afe9-4730-a1e0-1dbf56605fca",
   "metadata": {},
   "source": [
    "********************************************"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "515eb491-d47d-4f58-acc4-7de5a678cd7c",
   "metadata": {},
   "source": [
    "# 3. Copying your data to Jupyter <a class=\"anchor\" id=\"dataana\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95915561-043f-421f-962f-4ff66d96f77e",
   "metadata": {},
   "source": [
    "## What data do you need?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "149dc4d1-4b09-4829-944d-a5c089658ade",
   "metadata": {},
   "source": [
    "You need three pieces of information to run this eDNA analysis workflow.\n",
    "\n",
    "1. An amplicon abundance table, with samples as columns and taxa (ASVs) as rows\n",
    "2. A taxonomy table, where each ASV is linked to a taxonomic group (Kingdom ... Species) \n",
    "3. A samples table\n",
    "\n",
    "As long as you have these three pieces of information, you can run this Jupyter eDNA analysis workflow. Most upstream eDNA pipelines will generate this information, but this workflow has been specifically designed to use the output produced by the [anacapa pipeline](https://ucedna.com/software). If you have generated your data using a different upstream eDNA pipeline, this Jupyter workflow will need to be modified slightly to use it. Contact Paul Whatmore - paul.whatmore@qut.edu.au (the author of this workflow) if you need this or any other support."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0add81d1-2720-4ec3-854f-06609203886a",
   "metadata": {},
   "source": [
    "**Anacapa produces an output file that contains both 1. An amplicon abundance table and 2. A taxonomy table, combined together. So you will just need to upload this file to Jupyter and a samples table.**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf176ef7-d6b4-4c28-8fe2-878be501639f",
   "metadata": {},
   "source": [
    "Your anacapa output file will be in a subdirectory in your main anacapa directory, called 'xxxx_anacapa_output/xxxx/xxxx_taxonomy_tables/', where xxxx = the primer pair name you provided for anacapa (e.g. '16Smam'). The file itself is called 'xxxx__ASV_taxonomy_brief.txt'. Again, xxxx - primer pair name."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5001362a-185a-4400-a92e-f6b85c4183b3",
   "metadata": {},
   "source": [
    "You will also need a samples table. This needs to be in a specific format and structure. It needs a column called 'sample-id' and a column for each variable you have (such as location, collection date, etc). An example of the required structure is seen below. Create this in Excel and save it as 'sample_table' and as a csv file (`File` -> `Save as` -> drop down to select `CSV (Comma delimited) (*.csv)`).  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3feb4765-de27-4cfb-8753-2a4319309f42",
   "metadata": {},
   "source": [
    "| sample-id  | Species | Dummy1 | Dummy2 | Dummy3 |\n",
    "|------------|---------|--------|--------|--------|\n",
    "| Samp10.S10 | Owl     | AB     | Big    | 34     |\n",
    "| Sampl1.S1  | Owl     | AB     | Big    | 23     |\n",
    "| Sampl2.S2  | Owl     | AB     | Big    | 12     |\n",
    "| Sampl3.S3  | Owl     | CD     | Big    | 45     |\n",
    "| Sampl4.S4  | Owl     | EF     | Big    | 23     |\n",
    "| Sampl5.S5  | Owl     | CD     | Small  | 54     |\n",
    "| Sampl6.S6  | Owl     | CD     | Small  | 32     |\n",
    "| Sampl7.S7  | Owl     | CD     | Small  | 28     |\n",
    "| Sampl8.S8  | Owl     | EF     | Small  | 52     |\n",
    "| Sampl9.S9  | Owl     | EF     | Small  | 19     |"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f9df00a-b473-4349-a2de-0aa6937fec4d",
   "metadata": {},
   "source": [
    "## Uploading your data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "203109b5-c26b-4ccb-8b84-378fe7bdc516",
   "metadata": {},
   "source": [
    "First, create a folder called 'anacapa' in the left hand file panel in Jupyter. Righ click -> 'New folder'. Rename this to 'anacapa' (all lower case - R is case-sensitive). All your data files need to be in this folder and any files/figures created by this workflow will be output here. Open this folder."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49a0be5e-db1e-4abe-9074-2ed507224930",
   "metadata": {},
   "source": [
    "You can now drag and drop your two data files ('xxxx__ASV_taxonomy_brief.txt' and 'samples_table.csv') to this anacapa folder. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1a3f852-e693-4e94-8ee8-91fcf784bb94",
   "metadata": {},
   "source": [
    "*************************************"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74629662-6418-442e-8496-0cbe8bd328f7",
   "metadata": {},
   "source": [
    "# 4. Data overview <a class=\"anchor\" id=\"spana\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c653bc28-51bf-40a8-a919-a3511ed28b0a",
   "metadata": {},
   "source": [
    "As a basic example of how to examine your data in Jupyter, in this section we'll import your data file and display it as a table. This is also to check if you have the correct data files uploaded and they have the correct structure."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c1ed862-a4e9-4c48-afdc-8f250ca51497",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Amplicon abundance table\n",
    "\n",
    "The following table shows a count of the number of amplicons that mapped to a taxonomic group for each sample. Taxonomic groups are identified as ASVs - [Amplicon Sequence Variant](https://www.nature.com/articles/ismej2017119). This amplicon abundance table (absolute abundance) provides the foundation for the analysis in this analysis pipeline. Quality filtering was done by DADA2, which performs several sequence quality control steps: quality filtering, denoising, read pair merging and PCR chimera removal.\n",
    "\n",
    "These steps are outlined in detail in the DADA2 article (https://www.nature.com/articles/nmeth.3869):"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73e1f038-29ba-4de2-bec1-0acd3d5effc5",
   "metadata": {},
   "source": [
    "`DADA2, \"Callahan, Benjamin J., et al. \"DADA2: high-resolution sample inference from Illumina amplicon data.\" Nature methods 13.7 (2016): 581.\"`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d218374-c281-4d29-8613-a24a82916ecc",
   "metadata": {},
   "source": [
    "First, choose your working directory. This is the directory where your anacapa output files are (abundance table and samples table). See the previous section for details."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef7434f7-0683-4f00-b2b8-f6b34ea58c2f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "setwd(\"~/anacapa\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3258789-6d86-4e7f-a53b-11dd952a398b",
   "metadata": {},
   "source": [
    "Now, provide an ID for your project. This must be the same name as in your anacapa output file (see previous '3. Copying your data to Jupyter' section). So in the 'xxxx_ASV_taxonomy_brief.txt' file you uploaded, your project ID will be 'xxxx'. E.g. if your uploaded file is called '16Smam_ASV_taxonomy_brief.txt' your project ID will be called '16Smam'. Edit the below code cell to match your anacapa output file name."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b93337fd-e96b-459f-9993-843cdf0ffe1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "project_id <- \"rbcl\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8291573a-2d08-41b7-a4f0-b4b80091c702",
   "metadata": {},
   "source": [
    "Then, we import the abundance table into R:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "563f73d1-357c-4344-a1e6-b058804f1bb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "asvtable <- read.table(paste0(\"./\" , project_id, \"_ASV_taxonomy_brief.txt\"), check.names = FALSE, sep = \"\\t\", stringsAsFactors = FALSE, comment.char = \"\", header = T)\n",
    "colnames(asvtable)[1] <- \"ASV\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8cf64adc-91e9-4eb5-97e6-33ef3ee4256c",
   "metadata": {},
   "source": [
    "<mark><font color=\"red\">Pia, I'm also removing some repeated text from your sample names, as they appear on the ASV table (named after the fastq files). This is specific to your owl test dataset only and won't work on other datasets:</font></mark>\n",
    "\n",
    "<mark><font color=\"red\">Dataset1:</font></mark>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a775f743-38c4-4607-8c8d-d6abf48ab259",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(asvtable) <- gsub(\"X16Smam_Pia357NanoTest\", \"\", colnames(asvtable))\n",
    "colnames(asvtable) <- gsub(\".L001\", \"\", colnames(asvtable))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc7615b5-dd7d-4362-892b-30728b36667b",
   "metadata": {},
   "source": [
    "<mark><font color=\"red\">\n",
    "Dataset2:</font></mark>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "729f965b-8eba-41f5-b2f3-248f58507fbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(asvtable) <- gsub(\"cytbvert_\", \"\", colnames(asvtable))\n",
    "colnames(asvtable) <- gsub(\".L001\", \"\", colnames(asvtable))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d98a8a2d-723b-47b6-bee5-34e284dd2705",
   "metadata": {},
   "source": [
    "<mark><font color=\"red\">\n",
    "Plant data:</font></mark>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "345fb06e-d433-41a2-997b-1c9fe0921d2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(asvtable) <- gsub(\"rbcl_\", \"\", colnames(asvtable))\n",
    "colnames(asvtable) <- gsub(\"\\\\.\", \"_\", colnames(asvtable))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8c30fb6-11d9-43fd-97ab-e0aad6219d60",
   "metadata": {},
   "source": [
    "To generate the table we'll use the [DT: datatables package](https://rstudio.github.io/DT/). This creates a table that is searchable, columns can be ordered, table can be exported as a csv or Excel file, etc. Load the DT library:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fda7c446-0d61-41ae-8f78-a7b10c90750a",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(DT)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27355357-6716-44b4-a55c-a9d9cf5de232",
   "metadata": {},
   "source": [
    "Now generate the table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "518acb5e-2229-4b27-8b59-485bce06028f",
   "metadata": {},
   "outputs": [],
   "source": [
    "DT::datatable(asvtable, rownames = F,\n",
    "              width = \"100%\",\n",
    "              extensions = 'Buttons',\n",
    "              options = list(scrollX = TRUE,\n",
    "                             dom = 'Bfrtip',\n",
    "                             columnDefs = list(list(className = 'dt-center', targets=\"_all\")),\n",
    "                             buttons =\n",
    "          list('copy', 'print', list(\n",
    "            extend = 'collection',\n",
    "            buttons = list(\n",
    "                list(extend = 'csv', filename = \"ASV_table\"),\n",
    "                list(extend = 'excel', filename = \"ASV_table\"),\n",
    "                list(extend = 'pdf', filename = \"ASV_table\")),\n",
    "            text = 'Download'\n",
    "          ))\n",
    "      )\n",
    "    ) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a5b6ed2-82a8-4bba-9452-74ca53d24515",
   "metadata": {},
   "source": [
    "## Samples table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3e3b42f-8aba-48ee-88e8-77e245ce4bf1",
   "metadata": {},
   "source": [
    "Now import your samples table into R and view it as a table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6471928-1ef0-42c1-a43f-38ee22de793e",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples_table <- read.csv(\"sample_table.csv\", header = T)\n",
    "samples_table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9fb7bb4f-cd9d-4a8e-949e-8a3471dc90c6",
   "metadata": {},
   "source": [
    "**IMPORTANT: The sample IDs in the sample table must exactly match the sample IDs (column names) in the ASV table.** If they don't, you'll need to manually edit the sample names in the samples table and re-upload it."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3af787f6-f02a-4144-93e6-db70a63e11da",
   "metadata": {},
   "source": [
    "[Click here to go to the next section: 5. Filtering taxonomic assignments](./anacapa_5_filtration.ipynb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd9c26f0-4bc2-4d96-a876-00edba5b8b76",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
