{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1ab1952d-006a-4f33-8382-4b0557cd6847",
   "metadata": {},
   "source": [
    "# 6. Community structure"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7f4b330-c635-41b7-900a-403763e53fd9",
   "metadata": {},
   "source": [
    "**************************"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffb44551-214c-491c-9250-fdf4c9b45227",
   "metadata": {},
   "source": [
    "## Section contents\n",
    "[1. Overview](#overview)\n",
    "\n",
    "[2. Setting up your analysis environment](#envana2)\n",
    "\n",
    "[3. Preparing your data](#dataana2)\n",
    "\n",
    "[4. Taxonomic abundance bar plots](#abbarana)\n",
    "\n",
    "[5. Taxonomic abundance box and whisker plots](#abboxana)\n",
    "\n",
    "[6. Choosing a categorical variable to analyse](#varana2)\n",
    "\n",
    "[7. Community structure by variable](#commana)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41d72fe4-e025-4947-b21f-365af7e98654",
   "metadata": {},
   "source": [
    "****************************"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5530e27-80ee-4c65-9de6-0bf4adac1006",
   "metadata": {},
   "source": [
    "## Overview <a class=\"anchor\" id=\"overview\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6a1b5d2-65ef-4e52-9737-9df1150c5f9a",
   "metadata": {},
   "source": [
    "This section of the analysis workflow examines the community structure - i.e. the proportion, variance and abundance of taxonomic groups found in each sample and treatment group.\n",
    "\n",
    "**Amplicon sequence variants**\n",
    "\n",
    "Amplicon sequence variants (ASV) was referenced in the previous alpha diversity section but, as it is a key concept, will be discussed in more detail here. Taxonomic groups in this study are based on ASVs, inferred using the [**DADA2 software package**](https://www.nature.com/articles/nmeth.3869). \n",
    "\n",
    "**DADA2** infers sample sequences exactly and resolves differences of as little as 1 nucleotide.\n",
    "\n",
    "**ASV vs OTU**: Traditionally Operational Taxonomic Units (OTUs) have been used in 16S ampicon studies. More recently ASVs have been used, due their improved accuracy in identifying taxa, particularly genus and species. Basically, OTUs utilise a similarity clustering method to identify taxa, whereas ASV are generated by quantifying exact sequence matches to a sequence database and then statistically adjusting this using confidence thresholds. \n",
    "\n",
    "The OTU method typically can identify 97% similarity (with any accuracy) whereas the ASV method can identify even single base-pair differences. This enables a finer resolution of taxa down to the genus and species level. Note that there is increasing 'fuzziness' toward the lower taxonomic levels, as the diversity within some taxa is greater than the diversity between this and other taxa (in other words, even with ASV, not all taxa can be resolved to lower taxonomic levels and this highly depedant on the taxonomic group involved). See here for more details: https://www.zymoresearch.com/blogs/blog/microbiome-informatics-otu-vs-asv\n",
    "\n",
    "**Figures used in this section**\n",
    "\n",
    "Taxonomic differences between groups is presented in different ways in this section, including:\n",
    "\n",
    " - Stacked bar plots - To examine proportional taxonomic diversity per sample.\n",
    "\n",
    " - Heatmaps - Raw read counts in the abundance matrix are normalised (transformed to percentages). For categorical variables only. Samples are aggregated by variable (e.g. generation) and the most abundant taxon compared.\n",
    "\n",
    " - Box and whisker plot - For categorical variables only. As with heatmps, samples aggregated by variable and ordered by mean read abundance across all samples."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79f04bc2-5f50-4da0-aaf4-c22735b8258c",
   "metadata": {},
   "source": [
    "*************************"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ec883dc-1457-4c37-ae96-406420bf102f",
   "metadata": {},
   "source": [
    "## 2. Setting up your analysis environment <a class=\"anchor\" id=\"envana1\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c14f26d4-2677-4dfa-b5f1-d8013b9a8c14",
   "metadata": {},
   "source": [
    "Before you can begin your analysis you need to set up certain requirements, such as setting your working directory and installing/loading required R packges."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb244224-d49f-4060-8ac2-330d7f7755f6",
   "metadata": {},
   "source": [
    "### Set your working directory"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a33940e-f5dc-4076-8b36-9a575abcc1f3",
   "metadata": {},
   "source": [
    "Your [working directory](https://r-coder.com/working-directory-r/) in R is a base directory where R looks for your anacapa data files (and outputs files to)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6818fe3b-e8c5-496a-b1d8-cbabafde29ba",
   "metadata": {},
   "source": [
    "You will need to find the path of your anacapa results directory on the HPC and paste the location into the code cell below (setwd(\"~/Paste/Your/anacapa/Results/Directory/Path/Here\"))."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1147e53d-b339-4b13-834c-3ab661428819",
   "metadata": {},
   "outputs": [],
   "source": [
    "setwd(\"~/anacapa\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0be7c55-8e26-411b-8d0d-805021e588ca",
   "metadata": {},
   "source": [
    "### Install the required R packages"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0f9c5fe-37d0-4ab6-9821-bb70140470cc",
   "metadata": {},
   "source": [
    "R does most of its analysis using [functions](https://www.tutorialspoint.com/r/r_functions.htm). Some of these are built into base R, but many come as external [packages](https://r-pkgs.org/intro.html), which need to be installed and loaded into R."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42c6c396-2b78-44ce-8864-163141f29cdf",
   "metadata": {},
   "source": [
    "Load any required packages that have previously been installed using the [library()](https://www.tutorialspoint.com/r/r_packages.htm) function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c486a8e-f579-4a2b-ba65-6ad338ed2bb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(tidyverse)\n",
    "library(scales)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a625e62f-cd2b-4a6e-8a1e-7c3466454de6",
   "metadata": {},
   "source": [
    "Other packages need to be installed first."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1787760-d5fc-4a1e-b0d5-ae1fb7894d60",
   "metadata": {},
   "source": [
    "The key R analysis package used was ampvis2 (['Tools for visualising amplicon data'](https://madsalbertsen.github.io/ampvis2/)). Results from anacapa (ASV tables, taxonomy assignment, etc) is used as input for ampvis2 in this notebook. Install and load ampvis2 (this will take a few minutes as there are multiple dependent packages installed):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9f6111e-14e4-404e-aef2-5f81bb3db8b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "install.packages(\"remotes\", verbose = F)\n",
    "remotes::install_github(\"MadsAlbertsen/ampvis2\", quiet = T)\n",
    "library(ampvis2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f035381b-1d30-4979-8bb4-1499d1d088aa",
   "metadata": {},
   "source": [
    "Define a set of colours for plotting. Some of these plots have multiple clusters and it's difficult to find eougn contrasting colours to visually separate the clusters. I've developed a set of 25 colours that I've found contrast well, that we can use in the plots for this (and other) sections."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a92e4c16-728d-42b2-83ab-b48a50f539f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "c25 <- c(\n",
    "  \"dodgerblue2\", \"#E31A1C\", # red\n",
    "  \"green4\",\n",
    "  \"#6A3D9A\", # purple\n",
    "  \"#FF7F00\", # orange\n",
    "  \"black\", \"gold1\",\n",
    "  \"skyblue2\", \"#FB9A99\", # lt pink\n",
    "  \"palegreen2\",\n",
    "  \"#CAB2D6\", # lt purple\n",
    "  \"#FDBF6F\", # lt orange\n",
    "  \"gray70\", \"khaki2\",\n",
    "  \"maroon\", \"orchid1\", \"deeppink1\", \"blue1\", \"steelblue4\",\n",
    "  \"darkturquoise\", \"green1\", \"yellow4\", \"yellow3\",\n",
    "  \"darkorange4\", \"brown\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6009c383-3762-4ebf-ab65-30a25bb1a4ee",
   "metadata": {},
   "source": [
    "Set the default width and height for plots output on this Notebook. You can modify this as you prefer. Note that every plot in this Notebook is followed by code to output it as a file and this code defines width/height separately from the options below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "215d27c2-0104-4347-9472-14c717d2f9a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=12, repr.plot.height=8)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b3fcdd1-b022-4118-88a3-845d777eca3d",
   "metadata": {},
   "source": [
    "****************************"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee1b6fe0-5746-44c8-b923-f94011b9c1e7",
   "metadata": {},
   "source": [
    "## 3. Preparing your data <a class=\"anchor\" id=\"dataana1\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8330decf-ee5e-44f5-996d-cea9941c46e7",
   "metadata": {},
   "source": [
    "This ampvis2-based analysis requires, as input, an R object in ampvis2 format, which has a specific structure. At minumum this requires a samples table, an ASV count table and a taxonomy table, which are then combined into a single ampvis2 object."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "711a4a5b-f952-4f2d-8ed5-3ecf0d043817",
   "metadata": {},
   "source": [
    "First you'll need to provide an ID for your project. This must be the project ID you used in the filtration section. See the previous section for details."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d799a178-1394-499c-8dc6-c1fc1025609d",
   "metadata": {},
   "outputs": [],
   "source": [
    "project_id <- \"rbcl\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4eb9ee8c-99d5-4a07-a6d6-a1c586fc7575",
   "metadata": {},
   "source": [
    "### Import the samples table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a20167b3-1be1-47cd-9cb1-7e8c28340053",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples_table <- read.csv(\"sample_table.csv\", header = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "810ab946-3b37-4fa7-8880-810530ab0604",
   "metadata": {},
   "source": [
    "Have a look at your samples table and variables (metadata). In the previous filtration section we didn't use this information, but when examining diversity indices, etc, the metadata is critical."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e19da3a6-d17b-4e59-a1e7-e93f65cfec5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples_table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d046663-1ca3-455e-9add-83b0282bc26b",
   "metadata": {},
   "source": [
    "### Import the ASV abundance table and taxonomy table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d41eb9c6-fe5e-45f3-871f-b7099939bfce",
   "metadata": {},
   "source": [
    "**IMPORTANT: this is the FILTERED data that you exported at the end of the previous filtration section. You must have run that section once (only once is needed) for this and following sections to work**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39e14dae-32ad-471e-b432-e803a57b5854",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_data <- read.csv(paste0(project_id, \"_filtered_data.csv\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61772ef5-82c4-4eb6-9e73-d260dbd78b8f",
   "metadata": {},
   "source": [
    "Have a look at the top few rows of your data. The first 'ASV' column should contain the ASV IDs, the next columns are the samples, followed by the taxonomy levels."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10d92f5e-d089-4d21-a1d6-a9763513ed5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(filtered_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99d6332b-6a40-494a-b7c5-49cbae2b866f",
   "metadata": {},
   "source": [
    "### Create the ampvis2 database"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7c62867-d92e-4a42-a8c8-1492e581e440",
   "metadata": {},
   "source": [
    "Combine the samples data with the ASV table using amp_load(). This creates an ampvis2 database that can be used by ampvis2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0bc8961-c5b4-40bc-ab1a-c1c01f71b204",
   "metadata": {},
   "outputs": [],
   "source": [
    "ampvisdata <- amp_load(otutable = filtered_data,\n",
    "              metadata = samples_table)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "290cd12d-975a-4fa9-b681-0c6445e47e02",
   "metadata": {},
   "source": [
    "You can see a summary of your dataset by simply running the object name. This shows you the number of taxa identified, total/average/maximum/minimum number of reads per sample, etc. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "939658ab-89ac-4c39-96bb-256e8d26ddc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "ampvisdata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be78d315-2fd2-4bda-a841-8502c7999ed6",
   "metadata": {},
   "source": [
    "************************************"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9878a03f-cb89-4f8e-a2da-32aebd0374d8",
   "metadata": {},
   "source": [
    "## 4. Taxonomic abundance bar plots <a class=\"anchor\" id=\"abbarana\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81b45ef4-a691-4edc-a388-69904ac2da42",
   "metadata": {},
   "source": [
    "This section generates stacked bar plots showing the proportion of taxa per sample, for each taxonomic level"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0d4ee7b-1f7c-4914-a010-09bddd103fbe",
   "metadata": {},
   "source": [
    "First, you can output the raw (absolute) ASV abundance data as a csv file. You can add this as supplemental data to a manuscript or use it in other plotting or statistical programs. If you want the normalised (relative) abundance data, change `normalise = F` to `normalise = T` in the below code "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90c4a9a5-afc3-48c4-ad1d-e1a5964ab0e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "amp_export_otutable(ampvisdata, filename = paste0(\"ASVtable\"), normalise = F, sep = \",\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a76ed34-907f-4571-9ed3-7bac38caca9c",
   "metadata": {},
   "source": [
    "Choose the taxonomic level you want to plot (Choose from Phylum, Class, Order, Family, Genus, Species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d43e6675-1d96-4d78-b066-4e13ce72ea39",
   "metadata": {},
   "outputs": [],
   "source": [
    "taxgroup <- \"Order\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3138c40-b534-4920-a59f-ab86d120eec9",
   "metadata": {},
   "source": [
    "Generate the data for plotting. > 1% relative abundance only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09e56d81-516f-4464-a2b7-0c10c8d1cf53",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run amp_boxplot to generate aggregated data for the chosen taxonomic level, but output as a data object rather than plotting\n",
    "x <- amp_boxplot(ampvisdata, tax_aggregate = taxgroup, tax_empty = \"remove\", normalise = F)\n",
    "# Filter out any taxa with > 1% relative abundance\n",
    "# First calculate mean on aggregated taxa\n",
    "x2 <- aggregate(x$data$Abundance, list(x$data$Display), mean)\n",
    "# Select taxa only with abundance > 1%\n",
    "x3 <- x2[x2$x > 0,]\n",
    "# Pull out just those taxa in x$data\n",
    "x4 <- x$data[x$data$Display %in% x3$Group.1, ]\n",
    "# Remove duplicate rows (means already calculated, duplicates only cause bias)\n",
    "x5 <- x4[!duplicated(x4), ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0bc6566e-721a-4212-b70f-eb1ce2661c10",
   "metadata": {},
   "source": [
    "If you want your samples to be in a specific order on the plot (e.g. to cluster them by groups) then you need to re-order the levels, as you did in the previous sections. If you don't want to re-order your samples\n",
    "\n",
    "Check the current order of your samples:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e730e43-c707-4466-bb04-5eef6809e58b",
   "metadata": {},
   "outputs": [],
   "source": [
    "levels(factor(x5$Sample))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "499bf3d9-7ad4-4bff-96bd-02acea4fdb8e",
   "metadata": {},
   "source": [
    "Set the order of the samples how you like (all samples have to be inlcluded) then re-order them by running the below two code cells:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b9cd64e-ad8d-4891-a850-8e615614e3c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "lev <- c(\"PS.F1\", \"PS.F4\", \"PS.F7\", \"PS.F11\", \"PS.F12\", \"PS.F21\", \"PS.F24\", \"PS.N6\", \"PS.N8\", \"PS.N10\", \"PS.N11\", \"PS.N12\", \"PS.N13\", \"PS.N19\", \"PS.N22\", \"PS.N23\", \"PS.N24\", \"PS.F16\", \"PS.F18\", \"PS.F5\", \"PS.F9\", \"PS.F14\", \"PS.F15\", \"PS.F17\", \"PS.F22\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f5a5e43-dfb1-4977-af89-0cfd061a1c14",
   "metadata": {},
   "outputs": [],
   "source": [
    "x5$Sample <- factor(x5$Sample, levels = lev)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfd3b98a-1bbe-4f17-a095-8f5b47bb3ff7",
   "metadata": {},
   "source": [
    "Generate stacked bar plot. \n",
    "\n",
    "As with previous plots, you can modify colours (`scale_fill_manual()` is using the `c25` vector of colours we defined in section 2. \n",
    "\n",
    "You can change the `c25` colours in section 2 or add colours of your choice to the below code, e.g. `scale_fill_manual(values = c(\"red\", \"blue\", \"green\"))`)\n",
    "\n",
    "You can also change axis labels, text size and angle, [default theme](https://ggplot2.tidyverse.org/reference/ggtheme.html), etc.\n",
    "\n",
    "Feel free to add additional modifications to this or any other plot in this Notebook. Here is a good guide for doing this: http://r-statistics.co/Complete-Ggplot2-Tutorial-Part2-Customizing-Theme-With-R-Code.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4897a7f-4e68-40e6-9f95-2f98eb745471",
   "metadata": {},
   "outputs": [],
   "source": [
    "p <- ggplot(x5, aes(x = Sample, y = Abundance, fill = Display))\n",
    "p <- p + geom_bar(position=\"fill\", stat=\"identity\") +\n",
    "  labs(y = \"Abundance (%)\") +\n",
    "  theme_classic() +\n",
    "  scale_y_continuous(label = label_percent()) + \n",
    "  scale_fill_manual(values = c25) + \n",
    "  theme(text = element_text(size = 18), axis.text.x = element_text(angle = 90, size=12), axis.text.y = element_text(size=14)) + \n",
    "  labs(fill = taxgroup)\n",
    "p"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84966855-255f-47b9-8d25-4e9d7cd3e5cf",
   "metadata": {},
   "source": [
    "To plot a different taxonomic level, You can change `taxgroup <- \"Phylum\"` above to another taxonomic level, then re-run the code from that point."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37b5b923-16c9-469f-ab22-fd5457b0d61b",
   "metadata": {},
   "source": [
    "You can save your plot as a 300dpi (i.e. publication quality) tiff or pdf file. **These files can be found in your working directory.**\n",
    "\n",
    "**Tip:** you can adjust the width and height of the saved images by changing `width =` and `height =` in the code below (and every time ggsave appears in this workflow). Pdf files can be opened within Jupyter, so a good way to test a suitable width/height would be to save the image by running the pdf code below with the default 20cm width/height, then open the pdf file by clicking on it in the file browser panel (to the left of this notebook), then change the width/height and repeat this process as needed."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5097b0f6-c18f-40f5-bce9-cda2a53b1eb1",
   "metadata": {},
   "source": [
    "Export as a 300dpi tiff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee54c635-a12c-4093-b49c-055305812b56",
   "metadata": {},
   "outputs": [],
   "source": [
    "tiff_exp <- paste0(\"community_tax_abund_bar_plot_\", taxgroup, \".tiff\")\n",
    "ggsave(file = tiff_exp, dpi = 300, compression = \"lzw\", device = \"tiff\", plot = p, width = 20, height = 20, units = \"cm\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e67335f-19fe-4422-a63b-c393c88d42c1",
   "metadata": {},
   "source": [
    "Export as a pdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6070a4ff-d673-40d8-9490-77ce212c8908",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf_exp <- paste0(\"community_tax_abund_bar_plot_\", taxgroup, \".pdf\")\n",
    "ggsave(file = pdf_exp, device = \"pdf\", plot = p, width = 20, height = 20, units = \"cm\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe907939-6275-4878-bbd3-823ea1365db1",
   "metadata": {},
   "source": [
    "******************************"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e372f89d-2ba5-4970-a363-fa19d748d799",
   "metadata": {},
   "source": [
    "## 5. Taxonomic abundance box and whisker plots <a class=\"anchor\" id=\"abboxana\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6088fe8d-1c8c-4282-8069-c4199a0775c6",
   "metadata": {},
   "source": [
    "This section contains box and whisker plots for each taxonomic level (phylum … species).\n",
    "\n",
    "As with the bar plots in the previous section, these B&W plots show the relative abundance (in %) of taxa. But whereas heatmaps show the mean %, box and whisker plots show the interquartile range between samples ([max, min, median, quartiles](https://courses.lumenlearning.com/fscj-introstats1/chapter/box-plots/)), thus providing additional statistical information."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "474f49a4-4678-48f4-98dd-1b0f29850917",
   "metadata": {},
   "source": [
    "Choose the taxonomic level you want to plot (Choose from Phylum, Class, Order, Family, Genus, Species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89119af9-48c4-450b-b1b9-e761893edb26",
   "metadata": {},
   "outputs": [],
   "source": [
    "taxgroup <- \"Genus\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88afb8bc-b5c3-4d64-85fa-9d1d60b5d237",
   "metadata": {},
   "source": [
    "Generate the plot (and modify the attributes as you like)\n",
    "\n",
    "Note the `tax_show = 10` attribute. This says to show the top 10 taxa. Change this as desired. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7de958c-4946-400f-ab74-13d04980a1df",
   "metadata": {},
   "outputs": [],
   "source": [
    "p <- amp_boxplot(ampvisdata, tax_aggregate = taxgroup, tax_empty = \"remove\", tax_show = 10, normalise = F) +\n",
    "theme_bw() +\n",
    "theme(text = element_text(size = 18), axis.text.x = element_text(size=16), axis.text.y = element_text(size=14)) +\n",
    "labs(x = taxgroup)\n",
    "p"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2feed7e0-c8c0-4d65-afc2-62817eb1f877",
   "metadata": {},
   "source": [
    "Export as a 300dpi tiff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c354e1c-44dc-479b-af2f-118831f02403",
   "metadata": {},
   "outputs": [],
   "source": [
    "tiff_exp <- paste0(\"community_tax_abund_BW_plot_\", taxgroup, \".tiff\")\n",
    "ggsave(file = tiff_exp, dpi = 300, compression = \"lzw\", device = \"tiff\", plot = p, width = 20, height = 20, units = \"cm\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0eadb328-f884-4507-a1d2-fb67c11a87bd",
   "metadata": {},
   "source": [
    "Export as a pdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b38007d9-4a0b-49e5-af47-327d86e7725e",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf_exp <- paste0(\"community_tax_abund_BW_plot_\", taxgroup, \".pdf\")\n",
    "ggsave(file = pdf_exp, device = \"pdf\", plot = p, width = 20, height = 20, units = \"cm\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc48f0e9-ba93-468a-be1f-bc044e3cdaf1",
   "metadata": {},
   "source": [
    "***********************************"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b10a2ac-d05b-4750-a56c-fb36146ceaa0",
   "metadata": {},
   "source": [
    "## 6. Choosing a categorical variable to analyse <a class=\"anchor\" id=\"varana2\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9aecd0c2-3324-4f71-bc50-7c8d097c4b9e",
   "metadata": {},
   "source": [
    "In your metadata you'll usually have multiple variables. These need to be analysed individually, by selecting the variable in this section, then running the remaining analysis sections on this chosen variable. You can then re-run the analysis on another variable by returning to this section, changing the variable name, then running again the remaining analysis sections."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "268c1056-d146-4db8-a645-092fdd9dfdb6",
   "metadata": {},
   "source": [
    "**NOTE** This section is for choosing categorical variables only. See section 8 onward for analysis of continous (i.e. numeric) variables."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49bba1b3-7f28-425e-8933-b3a6bb6c2bca",
   "metadata": {},
   "source": [
    "You can view your variables as column names in your samples_table:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21d131f3-21fb-47bc-9d60-eb25bb6d38c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(samples_table)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fba951a2-9f0b-4682-896d-f2bda02dfe72",
   "metadata": {},
   "source": [
    "Enter the column name of the variable you want to analyse (i.e. change `group <- \"Myvariable\"` in the below cell to your chosen variable's column name). This has to be exactly the same as the column name, including capitalisation, characters such as underscores, etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98f27da3-ea42-4b92-92af-9eba01fefb0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "group <- \"Site\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69fd6cdd-411f-4615-9ede-044ddb51f7b4",
   "metadata": {},
   "source": [
    "### Ordering your variable"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2df45243-8e21-4392-a3cc-26c9017b0a8d",
   "metadata": {},
   "source": [
    "The plotting done in ampvis2 is done by the [ggplot2](https://ggplot2.tidyverse.org/) package. ggplot [factorises](https://www.datamentor.io/r-programming/factor/) variables and automatically orders them on the plot by alphabetical order."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae039ef6-e7ee-4b54-9836-8952c6d73762",
   "metadata": {},
   "source": [
    "This can cause your groups to be ordered incorrectly on the plot axes (e.g. a time series may not be plotted sequentially). "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7538f495-fb94-4ac9-8ff9-770e1ab9affc",
   "metadata": {},
   "source": [
    "You can manually set the order of your variable here. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd3a445d-0e81-48b5-a7b9-7746052a832e",
   "metadata": {},
   "source": [
    "First have a look at how ggplot will order your variable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e53d8c81-ff97-4369-981d-119448388982",
   "metadata": {},
   "outputs": [],
   "source": [
    "levels(factor(ampvisdata$metadata[[group]]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ec17ae3-2e09-4b77-883d-7e2d086ccfe6",
   "metadata": {},
   "source": [
    "If these are in the order you want to see them on your plot axes, nothing needs to be done. If they are in the wrong order you need to order them manually by setting the [**levels**](https://www.datamentor.io/r-programming/factor/)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d50fbaa-a20d-4533-8f58-c0ef18fa0e30",
   "metadata": {},
   "source": [
    "Choose how you want to order your groups here:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d669bd9-b030-4e7f-adaf-7bb789101b73",
   "metadata": {},
   "outputs": [],
   "source": [
    "lev <- c(\"S1\", \"S2\", \"S3\", \"S4\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dac9b3cb-9f2a-41c9-9850-dcd2647f96c8",
   "metadata": {},
   "source": [
    "To order your variable you need to put **all** the variable levels into the `lev = c(..)`. Make sure each level is in double quotes and separated by a comma."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7bf603eb-6a37-4a45-8b1b-b016b80da422",
   "metadata": {},
   "source": [
    "Then run the following to apply the levels to your data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08e7bd6e-b850-44dc-bd77-25d7f2ebf1c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "ampvisdata$metadata[[group]] <- factor(ampvisdata$metadata[[group]], levels = lev)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b6d60d1-c31d-40a2-89ef-0a819d01e910",
   "metadata": {},
   "source": [
    "**************************************"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07878962-9b85-4765-bbcb-9de17ecd4a7f",
   "metadata": {},
   "source": [
    "## 7. Community structure by variable <a class=\"anchor\" id=\"commana\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3deee16-b69e-46d1-b865-9a7202495412",
   "metadata": {},
   "source": [
    "In this section you can generate heatmaps and box and whisker plots for the variable you selected above, and for each taxonomic level."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "253b1f92-244e-47d4-8e00-b62972772e61",
   "metadata": {},
   "source": [
    "If you want to analyse another variable here, go back to the '7. Choosing a categorical variable to analyse' section, choose another variable and re-run the Notebook from that point."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9144f53f-660f-4ad9-8040-2ca6a840a3b2",
   "metadata": {},
   "source": [
    "Choose the taxonomic level you want to plot (Choose from Phylum, Class, Order, Family, Genus, Species). This applies to both the heatmaps and B&W plots below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dcdcfd77-1b12-464e-b43b-1da4ab2e8af5",
   "metadata": {},
   "outputs": [],
   "source": [
    "taxgroup <- \"Family\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3923fa5b-774d-4e9d-bace-f9fe50e297c9",
   "metadata": {},
   "source": [
    "You can add a secondary variable here (e.g. `var2 <- \"Phase\"`), which will split the plots by that variable. If you don't want to examine a secondary variable, leave the below as `var2 <- NULL`. **You must run the below code cell regardless**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e331113-ba46-452d-9791-e00bdddaadc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "var2 <- \"Animal\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1097a771-ae59-4104-b11b-a1d901cdb45e",
   "metadata": {},
   "source": [
    "### Heatmaps"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5faeff5-f929-4a4c-992c-7ba85d5c78d1",
   "metadata": {},
   "source": [
    "This section contains heatmaps for each taxonomic level (class - species) for your selected variable. The number in each cell is the relative proportion (in %) of taxa per group."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07bf654a-532e-4e5f-92de-fd4de5c5ebdf",
   "metadata": {},
   "source": [
    "Generate the plot. There are a variety of attributes you can modify:\n",
    "\n",
    "Note the `tax_show = 5` attribute. This says to show the top 5 taxa. \n",
    "\n",
    "The `plot_values_size =` defines the size of the text in the heatmap cells.\n",
    "\n",
    "The `tax_add = NULL` adds an additional taxonomic grup to the taxa names (e.g. changing this to `tax_add = \"Phylum\"` and plotting Genus will name the taxa as 'phylum:genus' but leaving it as `tax_add = NULL` will just give the genus name.\n",
    "\n",
    "`showRemainingTaxa = T` will aggregate, in a single row, the remaining taxa not shown on this heatmap. Change to `F` if you don't want to see this.\n",
    "\n",
    "`color_vector = NULL` uses the default colour range (orange -> blue). You can change this by providing your own colour range, e.g. `color_vector = c(\"red\", \"green\")`. You can choose from the huge number of R colours here: http://www.stat.columbia.edu/~tzheng/files/Rcolor.pdf\n",
    "\n",
    "`facet_by = var2` splits the plot by the seconday variable you provided above"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "894927d2-8526-42eb-aadf-41109a2e5d42",
   "metadata": {},
   "outputs": [],
   "source": [
    "p <- amp_heatmap(ampvisdata, tax_show = 5, plot_values_size = 5, tax_add = NULL, showRemainingTaxa = T, color_vector = NULL, facet_by = var2, group_by = group, tax_aggregate = taxgroup, tax_empty = \"remove\")\n",
    "p"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40b96082-a4e9-44b3-abb0-0e4a20c11bd2",
   "metadata": {},
   "source": [
    "Add some additional plot modifications. Change or remove these as desired ([or add your own modifications](http://r-statistics.co/Complete-Ggplot2-Tutorial-Part2-Customizing-Theme-With-R-Code.html))."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ebf5a94-466b-4adb-8547-9fe0df8b95dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "p <- p +\n",
    "theme_bw() +\n",
    "theme(text = element_text(size = 18), axis.text.x = element_text(angle = 90, size=16), axis.text.y = element_text(size=14)) +\n",
    "labs(y = taxgroup)\n",
    "p"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb1231f0-ffd5-48b9-b9f8-235deef165e5",
   "metadata": {},
   "source": [
    "Export as a 300dpi tiff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e66b82c-7d6f-4f1e-8e02-c06b65247427",
   "metadata": {},
   "outputs": [],
   "source": [
    "tiff_exp <- paste0(\"community_tax_abund_heatmap_\", group, \"_\", taxgroup, \".tiff\")\n",
    "ggsave(file = tiff_exp, dpi = 300, compression = \"lzw\", device = \"tiff\", plot = p, width = 20, height = 20, units = \"cm\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5bfb9952-cff6-4a9a-929e-6b0d00210087",
   "metadata": {},
   "source": [
    "Export as a pdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5221fc28-1515-45a9-af78-4bcdf5600341",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf_exp <- paste0(\"community_tax_abund_heatmap_\", group, \"_\", taxgroup, \".pdf\")\n",
    "ggsave(file = pdf_exp, device = \"pdf\", plot = p, width = 20, height = 20, units = \"cm\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0483328-bf9a-4133-a72e-fdb46481dbe0",
   "metadata": {},
   "source": [
    "### Box and whisker plots"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5a0c1dd-33d5-44cb-90e0-2e0dd38fd7c4",
   "metadata": {},
   "source": [
    "This section contains B&W plots for each taxonomic level (class - species) for your selected variable. This is similar to the heatmaps you generated in section 6, but separates the B&W into groups per variable, thus enabling an examination of how the variable groups differ in terms of taxonomic abundance."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f12c1b9a-9c6f-41fa-9eac-6f1b45c04284",
   "metadata": {},
   "source": [
    "Generate the plot. There are a variety of attributes you can modify:\n",
    "\n",
    "Note the `tax_show = 5` attribute. This says to show the top 7 taxa (**Tip:** to look at specific taxa you can provide a vector of taxa names, e.g. `tax_show = c(\"Muridae\", \"Dasyuridae\")`).\n",
    "\n",
    "The `tax_add = NULL` adds an additional taxonomic grup to the taxa names (e.g. changing this to `tax_add = \"Phylum\"` and plotting Genus will name the taxa as 'phylum:genus' but leaving it as `tax_add = NULL` will just give the genus name."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec61ce1e-55e7-484b-a2da-e5d81d9cc866",
   "metadata": {},
   "outputs": [],
   "source": [
    "p <- amp_boxplot(ampvisdata, tax_show = 5, tax_add = NULL, tax_aggregate = taxgroup, tax_empty = \"remove\", group_by = group)\n",
    "p"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ed9bebd-9871-423c-9eb5-6a43478ef5dc",
   "metadata": {},
   "source": [
    "You can change additional properties of the plot here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1acdf48d-598e-4445-bfb6-460f6c426c6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "p$mapping$fill <- as.name(\".Group\")\n",
    "p <- p + theme_bw() +\n",
    "scale_color_manual(values=c25) + scale_fill_manual(values=c25) +\n",
    "labs(fill = group, x = taxgroup) +\n",
    "theme(text = element_text(size = 18), axis.text.x = element_text(size=18), axis.text.y = element_text(size=16)) +\n",
    "guides(color = \"none\")\n",
    "p"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f86b08b7-1526-40c7-a88d-c110a0b49099",
   "metadata": {},
   "source": [
    "Export as a 300dpi tiff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3a28afa-8ac1-41b0-9885-beba7c811827",
   "metadata": {},
   "outputs": [],
   "source": [
    "tiff_exp <- paste0(\"community_tax_abund_BW_\", group, \"_\", taxgroup, \".tiff\")\n",
    "ggsave(file = tiff_exp, dpi = 300, compression = \"lzw\", device = \"tiff\", plot = p, width = 20, height = 20, units = \"cm\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ec09d28-d7cf-4028-af93-9df708788b66",
   "metadata": {},
   "source": [
    "Export as a pdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87ebbe0e-6cac-466a-837a-76d79d37bc13",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf_exp <- paste0(\"community_tax_abund_BW_\", group, \"_\", taxgroup, \".pdf\")\n",
    "ggsave(file = pdf_exp, device = \"pdf\", plot = p, width = 20, height = 20, units = \"cm\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59910eb5-26be-47d8-85f3-070474f5299c",
   "metadata": {},
   "source": [
    "**************************************"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "557feae4-d48d-4aac-876c-bc9f5472ed32",
   "metadata": {},
   "source": [
    "[Click here to go to the next section: 9. Differential abundance](./anacapa_9_DA.ipynb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7570b317-111b-4a60-a2dd-ded75192c84b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
